Add parameters component

ivanpauno opened the pull request:
# 🎉 New feature

Replaces https://github.com/ignitionrobotics/ign-gazebo/pull/1280.
Same as https://github.com/ignitionrobotics/ign-transport/pull/304 but rebased to target `ignition-transport11`.

Connected to:
- https://github.com/ignitionrobotics/ign-gazebo/pull/1431
- https://github.com/ignitionrobotics/ign-msgs/pull/241

## Summary
<!--Explain changes made, the expected behavior, and provide any other additional
context (e.g., screenshots, gifs) if appropriate.-->

Adds an implementation of a parameter registry, which allows through services to:

- Declare a parameter.
- Get the value of a parameter.
- Set the value of a parameter.
- List all parameters.

The same operations can be done directly locally, using the parameters registry API.
There's also a parameters client implementation, which is just a wrapper on top of sending requests to the services.

Parameter types are defined as protobuf messages.
This PR also adds a `ign param` cli command, which allows to easily interact with parameters.

TODO:

- [ ] Add tests!!!
- [ ] Open PR in ignition gazebo using this!!!

## Test it
<!--Explain how reviewers can test this new feature manually.-->

TODO

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

caguero reviewed (changes requested):
See also the style errors reported by CI.

ivanpauno commented on parameters/include/ignition/transport/parameters/Registry.hh:
```diff
@@ -0,0 +1,108 @@
+          const std::string & _parameterName,
+          std::unique_ptr<google::protobuf::Message> _initialValue);
+      
+        public: ParameterValue GetParameter(
```
Just to clarify, this is not a "getter", it's allowing to get any parameter from the registry.
e.g.: `registry.GetParameter("my_parameter_name");`

Though `registry.Parameter("my_parameter_name");` would also be clear.

ivanpauno replied:
Done in https://github.com/ignitionrobotics/ign-transport/pull/305/commits/3f9f57c5bc7e71945c1fc15e81ee5cca2595076d

ivanpauno commented on parameters/include/ignition/transport/parameters/Client.hh:
```diff
@@ -0,0 +1,71 @@
+        DeclareParameter(const std::string & _parameterName, const google::protobuf::Message & _msg);
+
+        ignition::msgs::ParameterDeclarations
+        ListParameters();
```
That's possible, but I have to make the `ignition::transport::Node node` member `mutable` for that to work.
Is that fine?

ivanpauno replied:
See https://github.com/ignitionrobotics/ign-transport/pull/305/commits/1ab769c4a37f97f602975bb9b6134942edddb0fb

ivanpauno commented on parameters/include/ignition/transport/parameters/Registry.hh:
```diff
@@ -0,0 +1,108 @@
+
+      struct ParametersRegistryPrivate;
+
+      class ParametersRegistry
```
Done in https://github.com/ignitionrobotics/ign-transport/pull/305/commits/452ee75d703a816b26dbd20e258429113862cc33.

ivanpauno commented on parameters/include/ignition/transport/parameters/Registry.hh:
```diff
@@ -0,0 +1,108 @@
+
+#include <google/protobuf/message.h>
+
+#include "ignition/msgs/parameter_declarations.pb.h"
```
Done in https://github.com/ignitionrobotics/ign-transport/pull/305/commits/452ee75d703a816b26dbd20e258429113862cc33.

ivanpauno commented on parameters/include/ignition/transport/parameters/Registry.hh:
```diff
@@ -0,0 +1,108 @@
+          std::string protoType;
+        };
+
+        public: ParametersRegistry(std::string _parametersServicesNamespace);
```
Done in https://github.com/ignitionrobotics/ign-transport/pull/305/commits/76bd87ec9546f66029282956f345e4342b486d31.

ivanpauno commented on parameters/include/ignition/transport/parameters/Registry.hh:
```diff
@@ -0,0 +1,108 @@
+
+        public: void SetParameter(
+          const std::string & _parameterName,
+          google::protobuf::Message & _value);
```
Done in https://github.com/ignitionrobotics/ign-transport/pull/305/commits/0510e18cc9e7562e975de8a25956a9702b114307

ivanpauno commented on parameters/include/ignition/transport/parameters/Registry.hh:
```diff
@@ -0,0 +1,108 @@
+          std::unique_ptr<google::protobuf::Message> _initialValue);
+      
+        public: ParameterValue GetParameter(
+          const std::string & _parameterName);
```
Fixed in https://github.com/ignitionrobotics/ign-transport/pull/305/commits/0510e18cc9e7562e975de8a25956a9702b114307

ivanpauno commented on parameters/include/ignition/transport/parameters/Registry.hh:
```diff
@@ -0,0 +1,108 @@
+        public: ignition::msgs::ParameterDeclarations ListParameters();
+
+        public: template<typename ProtoMsgT>
+        ProtoMsgT GetParameter(const std::string & _parameterName)
```
Done, see https://github.com/ignitionrobotics/ign-transport/pull/305/commits/0510e18cc9e7562e975de8a25956a9702b114307

ivanpauno commented on parameters/include/ignition/transport/parameters/Client.hh:
```diff
@@ -0,0 +1,71 @@
+          unsigned int _timeoutMs = kDefaultTimeoutMs);
+
+        std::unique_ptr<google::protobuf::Message>
+        GetParameter(const std::string & _parameterName);
```
See discussion [here](https://github.com/ignitionrobotics/ign-transport/pull/305#discussion_r857902800)

ivanpauno commented on parameters/src/Client.cc:
```diff
@@ -0,0 +1,154 @@
+
+#include <memory>
+#include <string>
+#include <sstream>
```
Done https://github.com/ignitionrobotics/ign-transport/pull/305/commits/1ab769c4a37f97f602975bb9b6134942edddb0fb

ivanpauno commented on parameters/include/ignition/transport/parameters/Client.hh:
```diff
@@ -0,0 +1,71 @@
+          unsigned int _timeoutMs = kDefaultTimeoutMs);
+
+        std::unique_ptr<google::protobuf::Message>
+        GetParameter(const std::string & _parameterName);
```
See comment [here](https://github.com/ignitionrobotics/ign-transport/pull/305#discussion_r857714161), I will update both places with the resolution of that discussion.

ivanpauno commented on parameters/include/ignition/transport/parameters/Client.hh:
```diff
@@ -0,0 +1,71 @@
+        ListParameters();
+
+      private:
+        std::string serverNamespace;
```
Done, see https://github.com/ignitionrobotics/ign-transport/pull/305/commits/6489c73d578896ff3853b9f084bd439a7ef0c83b

ivanpauno commented:
@caguero could you take another look to this?
I think I've addressed all the previous comments.
I've also rebased the PR.

CI seems to be failing because https://github.com/gazebosim/gz-msgs/pull/241 is required first.

caguero reviewed (changes requested):
This is looking good to me. Are you planning to add any tests to cover the new functionality?

ivanpauno commented:
@caguero I've added unit tests to cover both `ParametersRegistry` and `ParametersClient`, and fixed some bugs.

I think I will take another look to both classes API and uniformize them (and maybe add a common base class to both), but those changes should be minor. If you have any input about that let me know.

caguero commented:
> @caguero I've added unit tests to cover both `ParametersRegistry` and `ParametersClient`, and fixed some bugs.
> 
> I think I will take another look to both classes API and uniformize them (and maybe add a common base class to both), but those changes should be minor. If you have any input about that let me know.

Thanks! Do you mind to update https://github.com/gazebosim/gz-msgs/pull/241 with the changes that @mjcarroll requested? We can merge that one first, and then, merge this PR.

ivanpauno commented:
> Thanks! Do you mind to update https://github.com/gazebosim/gz-msgs/pull/241 with the changes that @mjcarroll requested? We can merge that one first, and then, merge this PR.

Done

> I think I will take another look to both classes API and uniformize them (and maybe add a common base class to both), but those changes should be minor. If you have any input about that let me know.

Done as well

ivanpauno commented:
@caguero friendly ping

scpeters commented:
I've merged https://github.com/gazebosim/gz-msgs/pull/241 and made a prerelease and opened a draft PR #349 that is level with this branch but using a special branch name that allows it to use matching branches in gzdev and homebrew-simulation to use the gz-msgs prerelease. I think the CI should now be working for this PR.

We should still make a stable release of gz-msgs before merging this, but I wanted to get working CI to support this PR review.

scpeters commented:
the windows build (which doesn't need the `ci_matching_branch/*` trick) is not happy:

* [![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=ign_transport-pr-win&build=333)](https://build.osrfoundation.org/job/ign_transport-pr-win/333/) https://build.osrfoundation.org/job/ign_transport-pr-win/333/

~~~
parameters\src\Client.cc(159,43): error C2589: 'constant': illegal token on right side of '::' [C:\Jenkins\workspace\ign_transport-pr-win\ws\build\ignition-transport11\parameters\src\ignition-transport11-parameters.vcxproj]

parameters\src\Client.cc(159): error C2062: type 'unknown-type' unexpected [C:\Jenkins\workspace\ign_transport-pr-win\ws\build\ignition-transport11\parameters\src\ignition-transport11-parameters.vcxproj]

parameters\src\Client.cc(159,53): error C2143: syntax error: missing ';' before '{' [C:\Jenkins\workspace\ign_transport-pr-win\ws\build\ignition-transport11\parameters\src\ignition-transport11-parameters.vcxproj]

parameters\src\Client.cc(199,43): error C2589: 'constant': illegal token on right side of '::' [C:\Jenkins\workspace\ign_transport-pr-win\ws\build\ignition-transport11\parameters\src\ignition-transport11-parameters.vcxproj]

parameters\src\Client.cc(199): error C2062: type 'unknown-type' unexpected [C:\Jenkins\workspace\ign_transport-pr-win\ws\build\ignition-transport11\parameters\src\ignition-transport11-parameters.vcxproj]

parameters\src\Client.cc(199,53): error C2143: syntax error: missing ';' before '{' [C:\Jenkins\workspace\ign_transport-pr-win\ws\build\ignition-transport11\parameters\src\ignition-transport11-parameters.vcxproj]

~~~

scpeters commented:
homebrew build just has one test failure that looks unrelated:

* [![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=ignition_transport-ci-pr_any-homebrew-amd64&build=1363)](https://build.osrfoundation.org/job/ignition_transport-ci-pr_any-homebrew-amd64/1363/) https://build.osrfoundation.org/job/ignition_transport-ci-pr_any-homebrew-amd64/1363/

scpeters commented:
I think the ABI checker complaint is a false positive

* https://build.osrfoundation.org/job/ignition_transport-abichecker-any_to_any-ubuntu_auto-amd64/465/API_5fABI_20report/

ivanpauno commented:
> the windows build (which doesn't need the ci_matching_branch/* trick) is not happy:

That's happening because `winerror.h` has a macro named NO_ERROR, see [here](https://stackoverflow.com/questions/49237883/macro-no-error-conflict).
The same name is being used in the enum I created [here](https://github.com/gazebosim/gz-msgs/blob/61607f967e1aabe2e238e1f590dddc2ca7e59ad0/proto/ignition/msgs/parameter_error.proto#L30).
Should I rename that to SUCCESS?

I can also try to fix that by adding an `undef` in the correct place and reordering the includes, but I would prefer not doing that.

scpeters commented:
> > the windows build (which doesn't need the ci_matching_branch/* trick) is not happy:
> 
> That's happening because `winerror.h` has a macro named NO_ERROR, see [here](https://stackoverflow.com/questions/49237883/macro-no-error-conflict). The same name is being used in the enum I created [here](https://github.com/gazebosim/gz-msgs/blob/61607f967e1aabe2e238e1f590dddc2ca7e59ad0/proto/ignition/msgs/parameter_error.proto#L30). Should I rename that to SUCCESS?

`SUCCESS` sounds good to me. I can make another prerelease after that `gz-msgs` change is merged

ivanpauno commented:
Done, see https://github.com/gazebosim/gz-msgs/pull/298.

scpeters commented on parameters/include/ignition/transport/parameters/Client.hh:
```diff
@@ -0,0 +1,118 @@
+
+      /// \brief Allow to get, set, declare or list parameters
+      /// \brief in a remote registry.
+      class ParametersClient : public ParametersInterface
```
each of these classes defined in header files that is intended to be public and a visible symbol should use the `IGNITION_TRANSPORT_PARAMETERS_VISIBLE` macro from `ignition/transport/parameters/Export.hh`

similar to `log/*.hh` header files:

* https://github.com/gazebosim/gz-transport/blob/8083eae49d87a029f3031a9857c9a0310551cf31/log/include/ignition/transport/log/Batch.hh#L24-L41

ivanpauno replied:
Thanks for the help!
CI should pass now

ivanpauno commented:
Ah sorry, it seems there're still some compile errors on Windows, but now it's different.
I will check what's going on

ivanpauno commented:
Ok, now the windows build passed.
There're a bunch of build warnings though, but the only one that seems related to the PR should've been fixed by https://github.com/gazebosim/gz-transport/pull/305/commits/4089abfeaaba996fb65f9e392ffd2d4a4d2011b2.
cc @scpeters

scpeters commented:
> Ok, now the windows build passed.
> There're a bunch of build warnings though, but the only one that seems related to the PR should've been fixed by [4089abf](https://github.com/gazebosim/gz-transport/commit/4089abfeaaba996fb65f9e392ffd2d4a4d2011b2).
> cc @scpeters

nice work! I just pushed one small fix to some comments in https://github.com/gazebosim/gz-transport/pull/305/commits/1688b6303e16b09aa53ecd04335a8d9c7087e92d where a different file was referring to `NO_ERROR` instead of `SUCCESS`

I'll sync #349 with this branch to hopefully get more of the CI passing

ivanpauno commented:
> nice work! I just pushed one small fix to some comments in https://github.com/gazebosim/gz-transport/commit/1688b6303e16b09aa53ecd04335a8d9c7087e92d where a different file was referring to NO_ERROR instead of SUCCESS

I think those two are unrelated to the PR, as that's part of `transport/log`, maybe there's another symbol called `NO_ERROR` in ignition.

scpeters commented on parameters/src/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+
+ign_add_component(parameters SOURCES ${sources} GET_TARGET_NAME param_lib_target)
+
+message(WARNING "param_lib_target ${param_lib_target}")
```
let's remove this line or convert to a `STATUS` message

* [![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=ignition_transport-ci-pr_any-ubuntu_auto-amd64&build=947)](https://build.osrfoundation.org/job/ignition_transport-ci-pr_any-ubuntu_auto-amd64/947/) https://build.osrfoundation.org/job/ignition_transport-ci-pr_any-ubuntu_auto-amd64/947/

ivanpauno replied:
Fixed in https://github.com/gazebosim/gz-transport/pull/305/commits/2ea957df386dbeae9fa0dbe99df0ecc4c9a4d1e5.

scpeters commented:
CI is working with the help of https://github.com/gazebosim/gz-transport/pull/349 (except for the ABI checker, which is complaining about untouched functions)

@caguero I think this is ready for further review

caguero commented on parameters/include/ignition/transport/parameters/Interface.hh:
```diff
@@ -0,0 +1,120 @@
+        ///   same name was declared before.
+        /// \throw ParameterInvalidTypeException (can only happen in client)
+        ///   if the parameter server does not recognize the parameter type.
+        public: virtual void DeclareParameter(
```
Usually the convention that we have is to avoid throwing exceptions if it's possible to continue. Can we potentially return a boolean to report that the parameter declaration didn't succeed and avoid the exception?

ivanpauno replied:
The problem with a boolean is that you lose information, you only inform that an error happened but not which.
If that's the convention, I can apply it here as well

caguero commented on parameters/src/Client.cc:
```diff
@@ -0,0 +1,237 @@
+  unsigned int timeoutMs;
+};
+
+ParametersClient::~ParametersClient() = default;
```
Nit: Can we add function separators?

ivanpauno replied:
Added in https://github.com/gazebosim/gz-transport/pull/305/commits/ccb520f2c2c1ba0897b371625e78a5ba53d28a74

caguero commented on parameters/src/Client.cc:
```diff
@@ -0,0 +1,237 @@
+{}
+
+static msgs::ParameterValue
+GetParameterCommon(
```
Nit: Normally we use lowercase for static functions.

ivanpauno replied:
Fixed in https://github.com/gazebosim/gz-transport/pull/305/commits/dff201bc9373b920ce1c4b1d94e9ad6a030f72bb

caguero commented on parameters/src/Client_TEST.cc:
```diff
@@ -0,0 +1,206 @@
+  ParametersClient client;
+}
+
+TEST_F(ParametersClientTest, Parameter)
```
Nit: Use function separator.

ivanpauno replied:
Done in https://github.com/gazebosim/gz-transport/commit/ccb520f2c2c1ba0897b371625e78a5ba53d28a74.

caguero commented on parameters/src/Utils.hh:
```diff
@@ -0,0 +1,45 @@
+      // Inline bracket to help doxygen filtering.
+      inline namespace IGNITION_TRANSPORT_VERSION_NAMESPACE {
+
+      std::string add_ign_msgs_prefix(const std::string & ignType);
```
Nit: Convert to camelCase and document with doxygen.

ivanpauno replied:
Done in https://github.com/gazebosim/gz-transport/pull/305/commits/dff201bc9373b920ce1c4b1d94e9ad6a030f72bb.

caguero commented on parameters/src/Utils.hh:
```diff
@@ -0,0 +1,45 @@
+/*
```
I suggest to also create a unit test for this class.

ivanpauno replied:
Done in https://github.com/gazebosim/gz-transport/pull/305/commits/8adc376da2b6150f1e7d92a45b89cfefe413a7cd

scpeters commented on parameters/src/Utils.hh:
```diff
@@ -0,0 +1,52 @@
+      /// \param any Message to get the type.
+      /// \return A string with the ignition protobuf type,
+      ///   or nullopt if it fails.
+      std::optional<std::string> getIgnTypeFromAnyProto(
```
these functions need `IGNITION_TRANSPORT_PARAMETERS_VISIBLE` in order to be called by the unit test on windows:

* [![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=ign_transport-pr-win&build=361)](https://build.osrfoundation.org/job/ign_transport-pr-win/361/) https://build.osrfoundation.org/job/ign_transport-pr-win/361/

~~~
Utils_TEST.obj : error LNK2001: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl ignition::transport::parameters::v11::addIgnMsgsPrefix(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?addIgnMsgsPrefix@v11@parameters@transport@ignition@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBV56@@Z) [C:\Jenkins\workspace\ign_transport-pr-win\ws\build\ignition-transport11\parameters\src\UNIT_Utils_TEST.vcxproj]

Utils_TEST.obj : error LNK2001: unresolved external symbol "class std::optional<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > __cdecl ignition::transport::parameters::v11::getIgnTypeFromAnyProto(class google::protobuf::Any const &)" (?getIgnTypeFromAnyProto@v11@parameters@transport@ignition@@YA?AV?$optional@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@std@@AEBVAny@protobuf@google@@@Z) [C:\Jenkins\workspace\ign_transport-pr-win\ws\build\ignition-transport11\parameters\src\UNIT_Utils_TEST.vcxproj]
~~~

ivanpauno replied:
Thanks, fixed in https://github.com/gazebosim/gz-transport/pull/305/commits/b4bb900d3e686eda279d72fc4e0081b5688497cf

scpeters commented:
I just synced #349 with this branch, which should give a full round of CI with the prereleases

scpeters commented:
codecov shows 0% code coverage of ParamCommandAPI.cc 

* https://app.codecov.io/gh/gazebosim/gz-transport/pull/349

ivanpauno commented:
> codecov shows 0% code coverage of ParamCommandAPI.cc

Ok, I can add test cases for it.

ivanpauno commented:
> codecov shows 0% code coverage of ParamCommandAPI.cc

Added tests in https://github.com/gazebosim/gz-transport/pull/305/commits/d5c7c0ca833c7ef31f08cde81413c13841419592

ivanpauno commented:
@scpeters thanks, solved both issues in https://github.com/gazebosim/gz-transport/pull/305/commits/dc77e1d82cc67cf554ba90080e7144b82e14baa1 and https://github.com/gazebosim/gz-transport/pull/305/commits/34f61905cf190691acb18800bb413076e7b55edb

scpeters commented:
> @scpeters thanks, solved both issues in [dc77e1d](https://github.com/gazebosim/gz-transport/commit/dc77e1d82cc67cf554ba90080e7144b82e14baa1) and [34f6190](https://github.com/gazebosim/gz-transport/commit/34f61905cf190691acb18800bb413076e7b55edb)

hm...the macOS and Ubuntu Jammy builds are fine, but Focal and Bionic are failing to compile

~~~
parameters/src/cmd/ParamCommandAPI.hh:26:8: error: expected unqualified-id before string constant
     26 | extern "C" void cmdParametersList(const char * _ns);
        |        ^~~
~~~

ivanpauno commented:
> hm...the macOS and Ubuntu Jammy builds are fine, but Focal and Bionic are failing to compile

Sorry, my bad :man_facepalming:

scpeters commented:
CI is looking pretty good now.

* Windows has some pre-existing compiler warnings, otherwise clean
* the abi-checker is claiming an API break caused by [modified symbols](https://build.osrfoundation.org/job/ignition_transport-abichecker-any_to_any-ubuntu_auto-amd64/524/API_5fABI_20report/) in `HandlerStorage.hh`, which is not touched by this PR, so I think this is a false positive.
* the coverage checker complains that code coverage is at 83.97% for this diff instead of 88.88%

I think it's ready for another round of reviews by @caguero

scpeters commented:
I pushed to #349 so this CI is re-running

I'm also preparing a PR for a stable release of gz-msgs 8.7.0

scpeters commented:
> I'm also preparing a PR for a stable release of gz-msgs 8.7.0

I'll open the release PR after https://github.com/gazebosim/gz-msgs/pull/307

scpeters commented:
gz-msgs 8.7.0 has been released so I deleted the branch from #349

scpeters commented:
@osrf-jenkins run tests please

caguero reviewed (commented):
It looks good to me! My only concern is about the use of exceptions. I know that they have their advantages but the convention that we follow is to avoid them as much as possible and not to terminate the execution. Is there any chance we could avoid exceptions? Maybe return an error structure with extra information as opposed to a boolean for not loosing information?

ivanpauno commented:
> It looks good to me! My only concern is about the use of exceptions. I know that they have their advantages but the convention that we follow is to avoid them as much as possible and not to terminate the execution. Is there any chance we could avoid exceptions? Maybe return an error structure with extra information as opposed to a boolean for not loosing information?

Done, please take a look to the last few commits.

scpeters commented:
it looks like there are several test failures

ivanpauno commented:
> it looks like there are several test failures

Yes sorry, I'm figuring those out locally.

scpeters commented on parameters/src/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+ign_add_component(
+  parameters SOURCES ${sources} GET_TARGET_NAME param_lib_target)
+
+message(WARNING "ignition-utils${IGN_UTILS_VER}")
```
this line should be removed

scpeters commented on parameters/include/ignition/transport/parameters/errors.hh:
```diff
@@ -0,0 +1,83 @@
+
+          private: ParameterResultType errorType;
+          private: std::string paramName;
+          private: std::string paramType;
```
there are two windows compiler warnings from these two lines

~~~
C:\Jenkins\workspace\ign_transport-pr-win\ws\gz-transport\parameters\include\ignition/transport/parameters/errors.hh(73,32): warning C4251: 'ignition::transport::parameters::v11::ParameterResult::paramName': class 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>' needs to have dll-interface to be used by clients of class 'ignition::transport::parameters::v11::ParameterResult' (compiling source file C:\Jenkins\workspace\ign_transport-pr-win\ws\gz-transport\parameters\src\errors.cc) [C:\Jenkins\workspace\ign_transport-pr-win\ws\build\ignition-transport11\parameters\src\ignition-transport11-parameters.vcxproj]

C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\xstring(4905): message : see declaration of 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>' (compiling source file C:\Jenkins\workspace\ign_transport-pr-win\ws\gz-transport\parameters\src\errors.cc) [C:\Jenkins\workspace\ign_transport-pr-win\ws\build\ignition-transport11\parameters\src\ignition-transport11-parameters.vcxproj]

C:\Jenkins\workspace\ign_transport-pr-win\ws\gz-transport\parameters\include\ignition/transport/parameters/errors.hh(74,32): warning C4251: 'ignition::transport::parameters::v11::ParameterResult::paramType': class 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>' needs to have dll-interface to be used by clients of class 'ignition::transport::parameters::v11::ParameterResult' (compiling source file C:\Jenkins\workspace\ign_transport-pr-win\ws\gz-transport\parameters\src\errors.cc) [C:\Jenkins\workspace\ign_transport-pr-win\ws\build\ignition-transport11\parameters\src\ignition-transport11-parameters.vcxproj]
~~~

ivanpauno replied:
That's weird...
I think it can be solved if I move the methods definitions to the .cpp file, so I will try that.

scpeters replied:
we can suppress the warnings if they aren't easy to fix

ivanpauno replied:
solved

scpeters commented:
I think the ABI checker is a false positive, and the windows CI problems are from pre-existing protobuf compiler warnings

@caguero have all your concerns been addressed?

caguero reviewed (approved):
Looking good!
